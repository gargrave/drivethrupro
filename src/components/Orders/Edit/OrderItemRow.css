.App .order-item-row .order-item-row-item + .order-item-row-item {
  margin-left: 30px;
  margin-top: 5px;
}
