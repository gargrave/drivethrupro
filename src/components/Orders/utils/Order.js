/*
A simple utility class for keeping track of details relating to an order.

If an existing Order object is passed, its data will be copied into this one;
this is useful for editing existing orders while maintaining the data (particularly the ID).

If nothing is passed in, a new Order will be created with default values.
*/

// A simple "primary key" mock for each object (i.e. to mock the functionality of a DB)
let nextID = 1

export default class Order {
  constructor (otherOrder) {
    if (otherOrder) {
      this.id = otherOrder.id
      this.items = otherOrder.items
      this.totalItems = otherOrder.totalItems
      this.totalPrice = otherOrder.totalPrice
    } else {
      this.id = nextID++
      this.items = []
      this.totalItems = 0
      this.totalPrice = 0
    }
  }

  addItem (orderItem) {
    this.items.push(orderItem)
    this.totalPrice += orderItem.totalPrice
    this.totalItems += orderItem.quantity
  }

  removeItem (orderItem) {
    if (this.items.find(i => i.id === orderItem.id)) {
      this.items = this.items.filter(i => i.id !== orderItem.id)
      this.totalPrice -= orderItem.totalPrice
      this.totalItems -= orderItem.quantity
    }
  }
}
