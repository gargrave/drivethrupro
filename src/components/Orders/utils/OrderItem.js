/*
A simple utility class for keeping track of a single line item within an order.
*/
let nextID = 1

export default class OrderItem {
  constructor (name, quantity, size) {
    this.id = nextID++
    this.name = `${name} (${size.name})`
    this.quantity = quantity
    this.pricePerUnit = size.price
    this.totalPrice = this.quantity * this.pricePerUnit
  }
}
