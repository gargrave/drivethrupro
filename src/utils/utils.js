export function displayAsDollars (cents) {
  if (!Number.isInteger(cents)) {
    throw new TypeError('"cents" must be an Integer.')
  }

  const prefix = cents < 0 ? '-' : ''
  const absValue = Math.abs(cents * 0.01)
  return `${prefix}$${absValue.toFixed(2)}`
}
