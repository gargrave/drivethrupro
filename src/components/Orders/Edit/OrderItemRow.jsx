import React from 'react'
import { func, number, shape, string } from 'prop-types'

import { displayAsDollars } from '../../../utils/utils'

import './OrderItemRow.css'

const OrderItemRow = props => {
  return (
    <li className="order-item-row">
      <span className="order-item-row-item order-item-row-name">
        {props.item.name}
      </span>
      <span className="order-item-row-item order-item-row-quantity">
        {props.item.quantity} @ {displayAsDollars(props.item.pricePerUnit)}
      </span>
      <span className="order-item-row-item order-item-row-price">
        {displayAsDollars(props.item.pricePerUnit * props.item.quantity)}
      </span>
      <button className="order-item-row-item remove-item-button" onClick={props.handleRemoveItemClick}>
        Remove
      </button>
    </li>
  )
}

OrderItemRow.propTypes = {
  item: shape({
    name: string.isRequired,
    quantity: number.isRequired,
    pricePerUnit: number.isRequired
  }).isRequired,
  handleRemoveItemClick: func.isRequired
}

export default OrderItemRow
