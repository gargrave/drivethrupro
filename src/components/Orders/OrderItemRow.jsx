import React from 'react'
import { number, shape, string } from 'prop-types'

import { displayAsDollars } from '../../utils/utils'

const OrderItemRow = props => {
  return (
    <li>
      {props.item.name} | {props.item.quantity} | {displayAsDollars(props.item.totalPrice)}
    </li>
  )
}

OrderItemRow.propTypes = {
  item: shape({
    name: string.isRequired,
    quantity: number.isRequired,
    totalPrice: number.isRequired
  }).isRequired
}

export default OrderItemRow
