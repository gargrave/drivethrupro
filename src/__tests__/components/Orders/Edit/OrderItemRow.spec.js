import React from 'react'
import { shallow } from 'enzyme'

import { displayAsDollars } from '../../../../utils/utils'
import OrderItemRow from '../../../../components/Orders/Edit/OrderItemRow'

describe('OrderItemRow', () => {
  let props

  beforeEach(() => {
    props = {
      item: {
        name: 'Test Item',
        quantity: 2,
        pricePerUnit: 5
      },
      handleRemoveItemClick: jest.fn()
    }
  })

  test('renders correctly', () => {
    const component = shallow(<OrderItemRow {...props} />)
    expect(component).toMatchSnapshot()
  })

  test('calls the "complete" callback when the button is clicked', () => {
    const component = shallow(<OrderItemRow {...props} />)
    expect(props.handleRemoveItemClick.mock.calls.length).toEqual(0)
    component.find('.remove-item-button').simulate('click')
    expect(props.handleRemoveItemClick.mock.calls.length).toEqual(1)
  })

  test('renders the name of the item', () => {
    const component = shallow(<OrderItemRow {...props} />)
    const text = component.find('.order-item-row-name').text()
    expect(text.indexOf(props.item.name)).not.toEqual(-1)
  })

  test('renders the quantity and price-per-unit of the item', () => {
    const component = shallow(<OrderItemRow {...props} />)
    const text = component.find('.order-item-row-quantity').text()
    expect(text.indexOf(props.item.quantity)).not.toEqual(-1)
    expect(text.indexOf(props.item.pricePerUnit)).not.toEqual(-1)
  })

  test('renders the total price of the item', () => {
    const component = shallow(<OrderItemRow {...props} />)
    const text = component.find('.order-item-row-price').text()
    const expected = displayAsDollars(props.item.quantity * props.item.pricePerUnit)
    expect(text).toEqual(expected)
  })
})
