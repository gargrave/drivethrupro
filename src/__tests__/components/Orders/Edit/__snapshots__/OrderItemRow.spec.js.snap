// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderItemRow renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <li
    className="order-item-row"
>
    <span
        className="order-item-row-item order-item-row-name"
    >
        Test Item
    </span>
    <span
        className="order-item-row-item order-item-row-quantity"
    >
        2
         @ 
        $0.05
    </span>
    <span
        className="order-item-row-item order-item-row-price"
    >
        $0.10
    </span>
    <button
        className="order-item-row-item remove-item-button"
        onClick={[Function]}
    >
        Remove
    </button>
</li>,
  "nodes": Array [
    <li
      className="order-item-row"
>
      <span
            className="order-item-row-item order-item-row-name"
      >
            Test Item
      </span>
      <span
            className="order-item-row-item order-item-row-quantity"
      >
            2
             @ 
            $0.05
      </span>
      <span
            className="order-item-row-item order-item-row-price"
      >
            $0.10
      </span>
      <button
            className="order-item-row-item remove-item-button"
            onClick={[Function]}
      >
            Remove
      </button>
</li>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <OrderItemRow
        handleRemoveItemClick={[Function]}
        item={
                Object {
                        "name": "Test Item",
                        "pricePerUnit": 5,
                        "quantity": 2,
                      }
        }
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "handleRemoveItemClick": [Function],
          "item": Object {
            "name": "Test Item",
            "pricePerUnit": 5,
            "quantity": 2,
          },
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <li
          className="order-item-row"
>
          <span
                    className="order-item-row-item order-item-row-name"
          >
                    Test Item
          </span>
          <span
                    className="order-item-row-item order-item-row-quantity"
          >
                    2
                     @ 
                    $0.05
          </span>
          <span
                    className="order-item-row-item order-item-row-price"
          >
                    $0.10
          </span>
          <button
                    className="order-item-row-item remove-item-button"
                    onClick={[Function]}
          >
                    Remove
          </button>
</li>,
        "_debugID": 2,
        "_renderedOutput": <li
          className="order-item-row"
>
          <span
                    className="order-item-row-item order-item-row-name"
          >
                    Test Item
          </span>
          <span
                    className="order-item-row-item order-item-row-quantity"
          >
                    2
                     @ 
                    $0.05
          </span>
          <span
                    className="order-item-row-item order-item-row-price"
          >
                    $0.10
          </span>
          <button
                    className="order-item-row-item remove-item-button"
                    onClick={[Function]}
          >
                    Remove
          </button>
</li>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <OrderItemRow
    handleRemoveItemClick={[Function]}
    item={
        Object {
            "name": "Test Item",
            "pricePerUnit": 5,
            "quantity": 2,
          }
    }
/>,
}
`;
